app/models/liquid_layout.rb

```diff
@@ -2,4 +2,21 @@ class LiquidLayout < ActiveRecord::Base
   has_many :campaign_pages
   validates :title, presence: true, allow_blank: false
   validates :content, presence: true, allow_blank: false
+  validate :real_partials
+
+  def partial_names
+    LiquidTagFinder.new(content).partial_names
+  end
+
+  def partial_refs
+    LiquidTagFinder.new(content).partial_refs
+  end
+
+  def real_partials
+    partial_names.each do |name|
+      partial = LiquidPartial.find_by(title: name)
+      errors.add :content, "includes unknown partial '#{name}'" if partial.nil?
+    end
+  end
```

Comment:
:+1: for the validation.

You could tidy this up a wee bit. presently `real_partials` is doing a little too much. You could take a more functional approach and move the checking of partials to `LiquidPartial`:

```ruby
class LiquidPartial

  # Returns new array of partial names which are
  # absent from the database.
  #
  def self.missing_partials(names)
    names.reject{|partial| LiquidPartial.exists?(title: name) }
  end
end
```
Then the responsibility of `real_partials` (`validate_presence_of_partials` might be a better name?) becomes immediately evident, and now it can respect SRP.

```ruby
  def validate_presence_of_partials
    LiquidPartial.missing_partials(partial_name).each do |name|
      errors.add :content, "includes unknown partial '#{name}'"
    end
  end
```